### ISSUE TITLE
Security Concerns

### ISSUE THREAD

**Issue by NovaGL** [bespoken/bst#332]
I see this has been updated to use a shorter URL.
The problem I have is I don't want this app to be able to access the root directory via GET.

Is there anyway to customise it.
I just did a .htaccess to disallow access.

I would like more of a security focus as this is essentially a backdoor to your local network.

**Comment by jkelvie** [comment]
Hi, when you say access to root directory, just want to make sure I understand. Are you proxying an HTTP service that has directory indexing on? 

In terms of customizing it, we could allow for HTTP methods to be enabled or disabled via parameter. I'm open to other suggestions as well.

**Comment by NovaGL** [comment]
Hi @jkelvie 

Let me rephrase my question.

BST Proxy poses a security concern more so now with the shorter urls.

I want an option to disable get (Yes I know I wanted GET enabled, but there are circumstances where this is not a good idea).

Disallow access without a skill id.
- For example - https://myskill.bespoken.link is not secure.
- Anyone could guess it and then use your proxy.
- Maybe white-list certain sub directories and reject everything else?

Add accepted domains
- Only accepts requests that come from specific domains, eg allow alexa requests but not request that might be from a hacker.

Disable short URL
- The old URL format even though it was longer it was harder to identify.

**Comment by jkelvie** [comment]
For the creation of secure tunnels, I believe the best approach is to:
1) Add a flag to proxy `--secure`
2) The flag will require that any request include either a query-string or header with `secret-key`
3) The secret-key must match the one configured with the proxy (from .bst/config)

By allowing for the key to be passed in the header, it will hide the key from being visible in the URL (a weakness with the query-string).

Only users with the key and the URL will be able to use the proxy.

@NovaGL and @jperata would like to hear your comments on this. We can then turn this into a specific issue.

**Comment by jperata** [comment]
To clarify:
From 1
This will add security in the local server, remote server will still pass data to the local server but it will be rejected if the secret-key is not present
From 2
This means that setup for skills would include that query string since there's no way for Amazon to generate the secret-key in the header. 

I agree with the solution

**Comment by jkelvie** [comment]
@jperata Yes with regard to both points. 

And point 1 is a significant one. Thank you for pointing this out. I believe this approach is simpler and still accomplishes the goal security-wise.

With regard to point 2, that is why we need to continue to support the query-string approach.

**Comment by NovaGL** [comment]
I agree with the solution too, as long as it's in the url. 

Can I make one small change though. I think this should be on as default and people can choose to have it off. 

Safer that way.

Maybe after you have coded that on the second phase could you look into like a .htacess or something/

Like my point mentioned a white or blacklist would be good to prevent subdomains from being open to the web that shouldn't be.

You might have a test and production environment on the same pc in different folders. You might only want the /test folder accessible

I am happy with the processing on the local end as long as unauthorised actions are rejected that's all that matters.

**Comment by jperata** [comment]
@NovaGL we released a new beta version (1.0.7) it already includes the --secure flag to reject any request that doesn't have the bespoken-key in the url.

**Comment by NovaGL** [comment]
Everything gives me unauthorised request what am I doing wrong?

**Comment by jkelvie** [comment]
Closing this issue. We will address additional security items as their own issues.
